nnstreamer/tensor_filter/tensor_filter_nntrainer.cc

```diff
@@ -0,0 +1,369 @@
+  gst_tensors_info_copy(&outputTensorMeta, &prop->output_meta);
+
+  return 0;
+}
+
+const char *NNTrainer::getModelConfig() { return model_config; }
+
+int NNTrainer::validateTensor(const GstTensorsInfo *tensorInfo, bool is_input) {
+
+  nntrainer::TensorDim dim;
+  nntrainer_tensor_info_s info_s;
+  unsigned int order[3] = {1, 3, 2};
+
+  if (is_input)
+    dim = model->getInputDimension();
+  else
+    dim = model->getOutputDimension();
+
+  g_assert(tensorInfo->info[0].type == _NNS_FLOAT32);
+  info_s.rank = NUM_DIM;
+
+  for (unsigned int i = 0; i < NUM_DIM - 1; ++i) {
+    g_assert(tensorInfo->info[0].dimension[i] == dim.getDim()[order[i]]);
+    info_s.dims.push_back(dim.getDim()[order[i]]);
+  }
+
+  g_assert(tensorInfo->info[0].dimension[NUM_DIM - 1] == dim.getDim()[0]);
+  info_s.dims.push_back(dim.getDim()[0]);
+
+  if (is_input) {
+    input_tensor_info.push_back(info_s);
+  }
+
+  return 0;
+}
+
+int NNTrainer::loadModel() {
+#if (DBG)
+  gint64 start_time = g_get_real_time();
+#endif
+  gsize file_size;
+  gchar *content = nullptr;
+  GError *file_error = nullptr;
+
+  g_assert(model_config != nullptr);
+  if (!g_file_test(model_config, G_FILE_TEST_IS_REGULAR)) {
+    ml_loge("the file of model_config (%s) is not valid (not regular)\n",
+            model_config);
+    return -1;
+  }
+
+  if (!g_file_get_contents(model_config, &content, &file_size, &file_error)) {
+    ml_loge("Error reading model config!! - %s", file_error->message);
+    g_clear_error(&file_error);
+    return -2;
+  }
+
+  try {
+    model = new nntrainer::NeuralNetwork(false);
+    model->loadFromConfig(model_config);
+  } catch (...) {
+    ml_loge("Cannot load model from config\n");
+    model->finalize();
+    return -1;
+  }
+
+#if (DBG)
+  gint64 stop_time = g_get_real_time();
+  g_message("Model is loaded: %" G_GINT64_FORMAT, (stop_time - start_time));
+#endif
+  return 0;
+}
+
+int NNTrainer::getInputTensorDim(GstTensorsInfo *info) {
+  gst_tensors_info_copy(info, &inputTensorMeta);
+  return 0;
+}
+
+int NNTrainer::getOutputTensorDim(GstTensorsInfo *info) {
+  gst_tensors_info_copy(info, &outputTensorMeta);
+  return 0;
+}
+
+int NNTrainer::run(const GstTensorMemory *input, GstTensorMemory *output) {
+#if (DBG)
+  gint64 start_time = g_get_real_time();
+#endif
+  std::vector<nntrainer::Tensor *> output_tensors;
+  nntrainer::Tensor *out = new nntrainer::Tensor();
+
+  std::vector<std::int64_t> d = input_tensor_info[0].dims;
+  nntrainer::Tensor X =
+    nntrainer::Tensor(nntrainer::TensorDim(d[3], d[0], d[2], d[1]),
+                      static_cast<float *>(input[0].data));
+
+  output_tensors.push_back(out);
+  std::shared_ptr<const nntrainer::Tensor> o;
+
+  o = model->inference(X);
+  if (o == nullptr) {
+    model->finalize();
+    delete out;
+    return -1;
+  }
+
+  *out = *o;
+
+  output[0].data = out->getData();
+
+  outputTensorMap.insert(std::make_pair(output[0].data, output_tensors[0]));
```

Comment:
IMPORTANT: as the data in the output tensor is a shared_ptr, storing just the data ptr wont suffice. The shared_ptr can be freed in the background
need to store the shared_ptr itself so as to stop it from freeing.